1 struct structure et mise en place de larchitecture de base de lapplication aspnet core

gaetan-denis opened the pull request:
# Structure et mise en place de l'architecture de base de l'application ASP.NET Core (Closes #1)


## Contexte : 

Création de la structure de base d'une API RESTful utilisant dotnet 8 permettant la gestion par un admin d'utilisateurs, de posts et de domaines associés. 

## Modifications apportées :

### Architecture de projet :

- Initialisation du projet ASP.NET Core 8 Web API.

- Création de la structure de la base de données avec les entités et leurs relations.

- Mise en place des couches Domain, Services et Controllers pour gérer les utilisateurs et les posts.
---
### Services et Répertoires :

- Ajout des services pour chaque entité.

- Création des interfaces et implémentation des classes associées.

- Implémentation du PasswordManager pour gérer le hachage des mots de passe.
---
### DTOs (Data Transfer Objects) :

- Création des DTOs pour chaque entité, y compris les validations nécessaires.
---
### Contrôleurs :

- Création des contrôleurs avec les verbes HTTP appropriés (GET, POST, PUT, DELETE).

- Ajout de la gestion d'erreurs et validation des données dans les services et contrôleurs.
---
### Fonctionnalités :

- Pagination ajoutée aux contrôleurs

- Ajout de la fonctionnalité de suppression.

- Création d'un formulaire de création d'utilisateur via l'API.
---
### UI et Frontend :

- Mise en place de l'application React avec la configuration du routage (React Router).

- Création des pages Dashboard, Users, Posts  et Domains avec récupération des données via l'API.

- Amélioration de l'ergonomie des pages avec des boutons "Edit" et "Delete".
---
### Autres :

- Correction de bugs de pagination et d'importation d'éléments.

- Implémentation de la gestion des erreurs au niveau des services.

- Améliorations de la sécurité avec des méthodes pour se protéger contre le XSS et SQL injection.
---
### Testé :
- Tests manuels sur les différentes routes des API (création, mise à jour, suppression d’utilisateurs, posts et domaines) via Swagger.

Commit:
STRUCT 📁 Initial setup of ASP.NET Core 8 Web API

Commit:
CHORE🧹 Added Entity Framework Core and SQL Server package

Commit:
STRUCT 📁 Created Domain layer and added User entity class with necessary properties

Commit:
STRUCT í³ Added Domain entity class to Domain layer with required
properties

Commit:
STRUCT 📁 Added User entity class to Domain layer with required propertie

Commit:
Added Content to Domain Entity and User Entity

Commit:
Added Content to Domain Entity and User Entity

Commit:
STRUCT 📁 Added Post  entity class to Domain layer with required properties

Commit:
Added Content to Post  Entity

Commit:
STRUCT 📁 Changed project's architecture to align with initial structure

Commit:
STRUCT 📁 Added User Relation with Post Class

Commit:
STRUCT 📁 Added Domain  Relation with Post Class

Commit:
STRUCT 📁 Added Post  Relation with Post Class

Commit:
FIX🔧 Corrected relationship in Domain  entity class

Commit:
STYLE🎨 Corrected spelling in comment

Commit:
STRUCT 📁 Change structure of project

Commit:
BUILD🔧 Installed AutoMapper.Extensions.Microsoft.DependencyInjection package

Commit:
BUILD🔧 Installed Microsoft.AspNetCore.OpenApi package (v8.0.10)

Commit:
BUILD🔧 Installed Microsoft.EntityFrameworkCore.Design package (v9.0.0)

Commit:
BUILD🔧 Installed Microsoft.EntityFrameworkCore.SqlServer package (v9.0.0)

Commit:
FEAT🚀 Added constraints to User class and added Email field

Commit:
FEAT🚀 Added constraints to Post class

Commit:
FEAT🚀 Added constraints to Domain class

Commit:
FEAT🚀 Added constraints to PostDomain class

Commit:
FEAT🚀 Added DataContext with entity relationships for Users, Posts, Domains, and PostDomains

Commit:
Suppression du fichier API.sln

Commit:
FEAT🚀 Added User and Post services to the Services layer

Commit:
FEAT🚀 Added Domain services to the Services layer

Commit:
CHORE📝 Fixed typo in User entity

Commit:
STRUCT📁 Created IPasswordHasherService interface for password hashing and verification

Commit:
REFACTOR♻️ Renamed IPasswordHasherService interface to IPasswordManager for consistency

Commit:
FEAT🚀 Implemented PasswordManager service to hash and verify passwords

Commit:
UPDATE🚀 Added PasswordManager service and removed default weather app setup

Commit:
STRUCT 📁 Create 'Dtos' directory and add 'AddPostDto' class

Commit:
STRUCT 📁 Modify DB context and add connection string in appsettings.json

- Updated the DB context to align with new database schema.
- Added the connection string configuration to 'appsettings.json' for database connectivity.

Commit:
STRUCT 📁 Add 'PostUpdateDto' and create three 'User' DTOs

- Added 'PostUpdateDto' for handling post update data.
- Created three DTOs for 'User' to manage user-related data transfer.

Commit:
STRUCT 📁 Add constraints to existing DTOs and create 'Domain' entity DTO

Commit:
STRUCT 📁 Implement 'UserController' with endpoints to get user data

Commit:
FEAT🚀: Creates UserController with HTTP verbs (GET, POST, PUT, DELETE), adds repositories folder, and starts business logic in UserService

Commit:
FEAT🚀: Implements valid data transition for User data from application to database

Commit:
FEAT🚀: Creates DomainController

Commit:
DOCS📚: Refines comments for clear documentation in PostController methods

Commit:
FIX🔧 Fix namespace error in DomainController.cs and STRUCT 📁 Add EErrorType enum in Enum folder

Commit:
FIX🔧 Update ServiceResponse to use EErrorType for error management

Commit:
FIX🔧 Updated error handling in UserService to use EErrorType for better error management

Commit:
FEAT🚀 Modified controllers to integrate new error handling system

Commit:
FEAT🚀 Added IPostRepository and IDomainRepository interfaces

Commit:
FEAT🚀 Implemented PostRepository, DomainRepository, PostService, and DomainService logic; integrated HttpManager and revamped error handling in UserService, PostService, and DomainService

Commit:
REFACTOR♻️ Renamed 'Enum' folder to improve clarity and understanding

Commit:
UPDATE🚀: Integrated error handling and documentation in PostService.

Commit:
UPDATE🚀: Integrated error handling and documentation in DomainService.

Commit:
UPDATE🚀: Integrated error handling and documentation in UserService.

Commit:
UPDATE🚀: Implemented password manager in UserService and added User_Salt field in the database.

Commit:
FIX🔧: Resolve display issue for User entity creation and update values in database.

Commit:
FEAT🚀: Implemented BaseRepository interface to standardize repository operations.

Commit:
REFAC♻️: Implemented BaseRepository interface and refactored UserRepository, PostRepository, and DomainRepository accordingly. Also, resolved multiple console warnings.

Commit:
STRUCT📁: Created 'Pagination' class in Services directory.

Commit:
FEAT🚀 Implement pagination in UserController

Commit:
FEAT🚀 Implement pagination in PostController

Commit:
FEAT🚀 Implement pagination in DomainController

Commit:
REFACTOR♻️ Refactor controller methods to use HttpManager for standardized response handling

Commit:
STRUCT 📁 Created 'Validador' folder, added 'PayloadValidator' file, and implemented 'ProtectAgainstSqlInjections' method

Commit:
STYLE 🎨 Fixed spelling mistakes in a comment

Commit:
FEAT 🚀 Added 'ProtectAgainstXSS' method and renamed 'ProtectAgainstSqlInjections' to 'ProtectAgainstSQLI'

Commit:
FEAT 🚀 Added 'BlockTemporaryEmails' method and improved class comments for better understanding

Commit:
FEAT 🚀 Implemented 'ProtectAgainstSQLI', 'ProtectAgainstXSS', and 'BlockTemporaryEmails' methods in all controllers for input validation and security

Commit:
FEAT🚀: Initialize React app and configure CORS policy

- Set up the React frontend application.
- Configured CORS policy in the backend to allow requests from local frontend

Commit:
STRUCT 📁 Create folders for components, hooks, pages, and services

Commit:
CHORE🧹 Install react-router-dom and its types for TypeScript

Commit:
FEAT🚀 Add routing and pages for Dashboard, Users, and Posts

Commit:
STYLE🎨 Reset CSS to default styles

Commit:
CHORE🧹 Push modified assembly info files

Commit:
STRUCT 📁 Create 'types' folder and add User and Post interfaces

Commit:
FEAT🚀 Add fetchUsers function to fetch user data from API

Commit:
CHORE🧹 Push updated build and debug files for API project

Commit:
FEAT🚀 Make UsersPage functional and display user list

Commit:
FIX🔧 Fix import error in UsersPage

Commit:
FEAT🚀 Refactor ApiResponse to support generic types

Commit:
FEAT🚀 Add user_id to Post interface

Commit:
FEAT🚀 Improve UsersPage functionality and UI

Commit:
FEAT🚀 Create functional PostPage to display all posts

Commit:
STYLE🎨 Change title of the main page

Commit:
CHORE🧹 Push updated build and debug files for API project

Commit:
FEAT🚀 Add 'Edit' and 'Delete' buttons on UserPage

Commit:
STYLE🎨 Improve page layout for better ergonomics

Commit:
STYLE🎨 Add SCSS stylesheets for UserPage and PostPage

Commit:
FEAT🚀 Add deleteUser function in api.ts

Commit:
CHORE🧹 Push updated build and debug files for API project

Commit:
FEAT🚀 Update Dto and UserService to return user ID

Commit:
FEAT🚀 Update Dto and PostService to return Post ID

Commit:
FEAT🚀 Update Domain to return user ID & uniformize page names

Commit:
FIX🔧 Fix missing 'key' prop bug across all pages (user, post, domain, etc.)

Commit:
FIX🔧 Fix error related to accessing an incorrect property on domain object

Commit:
feat🚀 Add delete functionality for users, posts, and domains

Commit:
FIX🔧 Corrected issue with the delete functionality not affecting the server for users, posts, and domains.

Commit:
FEAT🚀 Added a registration form for creating new users.

Commit:
FEAT🚀 Adds user creation functionality via API

Commit:
UPDATE🚀 Add 'password' property to User interface

Commit:
UPDATE🚀 Add 'password' field to user creation form

Commit:
FEAT🚀 Add functionality for user creation

Commit:
FEAT🚀 Implement function to add a new post via API

Commit:
FEAT🚀 Create function to add a new post

Commit:
FEAT🚀 Create function to add a new domain

Commit:
FEAT🚀 Create function to update a domain

Commit:
FEAT🚀 Create function to update a post

Commit:
REVERT🔄 Reset to a stable state and re-implement add post functionality

Commit:
FEAT🚀 Add functionality to add a new domain

Commit:
FIX🔧 Fix errors in DomainPage

Commit:
FEAT🚀 Add functionality to update a domain

Commit:
FEAT🚀 Add functionality to update posts

Yekuuun commented:
## Commentaires PR : 

### Description PR : 

- Pour la description de la PR (pas hésite à etre + générique => compliqué dans ce cas ci car beacoup de choses mais on en parle demain car y a quelques oublis (améliorer la structure de la description)

- oubli des labels etc dans la PR.

### Code BACKEND: 

- Attention aux directives using répétitives => pas hésite à utiliser des global using
![image](https://github.com/user-attachments/assets/df4c5645-5c22-4874-9eba-94bf87fcb940)

- Quid une fonction + générique pour le validator plutot que de tester manuellement les champs des DTO passés dans les controllers.

![image](https://github.com/user-attachments/assets/317c47be-d3ef-484b-a08d-1249eebe591e)


il y a moyen d'avoir une fonction qui passe dans l'objet et check les champs de l'objet + application d'un filtre selon le type (string, int etc)

- Réponse OK => attention a son utilisation (renvoyer un code 200 dans les controllers d'office n'est pas top

![image](https://github.com/user-attachments/assets/333dbd4b-080c-43b6-8428-cec48e46445f)


Ici, pour l'update du user, il pourrait y avoir des erreurs tel qu'un conflit, unauthorized etc. => HTTp Response manager.


- Les interfaces pour les repositories ne sont pas obkgatoires (voir mon projet)
![image](https://github.com/user-attachments/assets/d07460c0-3a2a-45ae-8e86-01693205e383)

- QUID, utiliser des mappers pour les retours dans le services ?

![image](https://github.com/user-attachments/assets/83e20421-c1e8-403e-a677-fce99ce7e6a1)

+ simple pour les retours que de maneullement créer le DTO de retour.

ex : 

![image](https://github.com/user-attachments/assets/d890a77e-ddb4-4d2d-93fd-3703a119b603)

response.data = _mapper.Map<PostDto>(post);

- Attention a l'indentation et aux espaces dans le code. Bien vérifier que le code respecte une belle synthaxe de dev

![image](https://github.com/user-attachments/assets/97148a09-e14c-4c48-b15c-b5f225b6fae2)

Commit:
WIP🚧 Work in progress: User update functionality (not yet functional)

Commit:
REVERT🔄 Reset to a stable state

Commit:
CHORE🧹 Remove bin and obj folders and add them to .gitignore

Commit:
REVERT🔄 Reset project to a stable state

Commit:
REFACTOR♻️ Move controller usings to GlobalUsing.cs

Commit:
REFACTOR♻️ Move Dtos and Entities  usings to GlobalUsing.cs

Commit:
REFACTOR♻️ Move Repositories,Services and Validator  usings to GlobalUsing.cs

Commit:
STYLE🎨 Properly reindent code for better readability

Commit:
STYLE🎨 Fix indentation in PayloadValidator.cs

gaetan-denis commented:
## Commentaires P.R.
### Description P.R. : 
- [ ] description améliorée en essayant d'être moins lacunaire et plus concis, à voir si cela convient ainsi.
- [x] labels rajoutés dans la P.R.
![image](https://github.com/user-attachments/assets/5f02babb-3a76-4086-9aec-1feef9056991)

---
## Code Backend : 
- [x] directives using retirés de chaques fichiers et rajoutés dans GlobalUsing.cs 
![image](https://github.com/user-attachments/assets/6183d56f-1f03-44b7-97af-fe35b3ba85cc)
mis à part pour PasswordManager pour lequel j'ai jugé plus pertinent de les laisser dans le fichier même. (sécurité?)
![image](https://github.com/user-attachments/assets/3e4fa928-65f8-4bc1-bb45-f20622d8a04a)


- [x] fonction plus générique pour le validator implémentée : ValidateObject renvoit un message générique en cas d'attaque XSS ou d'injection SQL.
![image](https://github.com/user-attachments/assets/5913030c-1cec-443c-9caf-487a37392ee4)

- [ ] Réponse ok : 🚧 en cours

Implémentation d'erreurs dans le HttpManager ainsi que dans l'Enum, pour gérer les erreurs de manières moins génériques.

À faire :

- Renvoyer un code 204 en cas de suppression réussie
- Renvoyer un code 202 en cas de création réussie.

![image](https://github.com/user-attachments/assets/8f20122a-9556-4f72-911c-062b97fe20d8)

- [x] Mapper pour les retours de services : implémentations sur les trois services  effectuées et opérationnelles.

![image](https://github.com/user-attachments/assets/53d6d9cb-634a-4923-8f21-46899955fa35)

⚠️ Provoque actuellement des erreurs avec ma mise en place dans React sur l'action de Delete :

- Un clic sur le bouton d'un suppression d'un élément, supprime toute la liste d'élément sans procéder à l'effacement des données quelles qu'elles soient. 

**Avant**

![image](https://github.com/user-attachments/assets/9aeb0ecf-a894-43e2-8801-9e06619ef1ab)

**Après**

![image](https://github.com/user-attachments/assets/735a1843-9127-4dc2-a39d-801062cdd1d8)



- [x] Indentation : corrigée de façon générale dans tous les fichiers.
![image](https://github.com/user-attachments/assets/b52b437f-97f7-40b8-8701-88927ebe44b3)

---

Commit:
FEAT🚀 Implement ValidateObject function in PayloadValidator.cs for more generic error handling

Commit:
FEAT🚀 Implement AutoMapper mapping in PostService

Commit:
FEAT🚀 Implement AutoMapper mapping in UserService

Commit:
FEAT🚀 Implement AutoMapper mapping in DomainService

Commit:
FEAT🚀: Updates EErrortype to include new errors, modifies HttpManager accordingly, and documents EErrortype

Commit:
FIX🔧 Clarify error names and fix resulting bugs

Commit:
FEAT🚀 Add getControllerName, getEntityId, and createDeleteResponse in HttpManager

- Introduced new methods for more specific HTTP responses
- Not yet functional, work in progress

Commit:
FIX🔧 Make CreateDeleteResponse generic to fix bug

Commit:
CHORE🧹 Install and configure Serilog for logging

- Added Serilog to improve logging capabilities
- Configured logging to help with debugging and issue tracking

Commit:
FEAT🚀 Add GetByUsernameOrEmailAsync method to IUserRepository

Commit:
FEAT🚀 Add GetByUsernameOrEmailAsync method to UserRepository

Commit:
REFACTOR♻️ Replace var with explicit types in  UserService
